collector/tcpstat.go

```diff
@@ -0,0 +1,150 @@
+// +build !notcpstat
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	procTCPStat  = "/proc/net/tcp"
+	procTCP6Stat = "/proc/net/tcp6"
+)
+
+type TCPConnectionState int
+
+const (
+	TCP_ESTABLISHED TCPConnectionState = iota + 1
+	TCP_SYN_SENT
+	TCP_SYN_RECV
+	TCP_FIN_WAIT1
+	TCP_FIN_WAIT2
+	TCP_TIME_WAIT
+	TCP_CLOSE
+	TCP_CLOSE_WAIT
+	TCP_LAST_ACK
+	TCP_LISTEN
+	TCP_CLOSING
+)
+
+type tcpStatCollector struct {
+	config Config
+	metric *prometheus.GaugeVec
+}
+
+func init() {
+	Factories["tcpstat"] = NewTCPStatCollector
+}
+
+// NewTCPStatCollector takes a config struct and returns
+// a new Collector exposing network stats.
+func NewTCPStatCollector(config Config) (Collector, error) {
+	return &tcpStatCollector{
+		config: config,
+		metric: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: Namespace,
+				Name:      "tcp_connection_states",
+				Help:      "Number of connection states.",
+			},
+			[]string{"state"},
+		),
+	}, nil
+}
+
+func (c *tcpStatCollector) Update(ch chan<- prometheus.Metric) (err error) {
+	tcpStats, err := getTCPStats(procTCPStat)
+	if err != nil {
+		return fmt.Errorf("couldn't get tcpstats: %s", err)
+	}
+
+	// if enabled ipv6 system
+	if _, hasIPv6 := os.Stat(procTCP6Stat); hasIPv6 == nil {
```

Comment:
Sorry, I have added modification. (93d21b5)
Fix always error on not exists `/proc/net/tcp6`.